sycl/include/syclcompat/group_utils.hpp

```diff
@@ -0,0 +1,1269 @@
+
+    const int target_id = static_cast<int>(id) + distance;
+    if ((target_id >= 0) && (target_id < group_work_items)) {
+      output = _local_memory[static_cast<size_t>(target_id)];
+    }
+  }
+  /// Selects a value from a work-item at a given distance in the work-group
+  /// and stores the value in the output, using a wrapped index to handle
+  /// overflow.
+  ///
+  /// \tparam ItemT The work-item identifier type.
+  /// \param item The work-item identifier.
+  /// \param input The input data to be selected.
+  /// \param output The output where the selected data will be stored.
+  /// \param distance The number of work-items to look ahead in the
+  /// work-group.
+  template <typename ItemT>
+  __syclcompat_inline__ void select2(const ItemT &item, T input, T &output,
+                                     unsigned int distance = 1) {
+    auto g = item.get_group();
+    size_t id = g.get_local_linear_id();
+    _local_memory[id] = input;
+
+    sycl::group_barrier(g, sycl::memory_scope::work_group);
+
+    unsigned int offset = id + distance;
+    if (offset >= group_work_items)
+      offset -= group_work_items;
+
+    output = _local_memory[offset];
+  }
+  /// Performs a shuffle operation to move data to the right across the
+  /// work-items, shifting elements in a work-item array by one position to the
+  /// right.
+  ///
+  /// \tparam ElementsPerWorkItem The number of data elements per work-item.
+  /// \tparam ItemT The work-item identifier type.
+  /// \param item The work-item identifier.
+  /// \param input The input data to be shuffled.
+  /// \param output The array that will store the shuffle result.
+  template <int ElementsPerWorkItem, typename ItemT>
+  __syclcompat_inline__ void shuffle_right(const ItemT &item,
+                                           T (&input)[ElementsPerWorkItem],
+                                           T (&output)[ElementsPerWorkItem]) {
+    auto g = item.get_group();
+    size_t id = g.get_local_linear_id();
+    _local_memory[id] = input[ElementsPerWorkItem - 1];
+
+    sycl::group_barrier(g, sycl::memory_scope::work_group);
+
+#pragma unroll
+    for (int index = ElementsPerWorkItem - 1; index > 0; --index)
+      output[index] = input[index - 1];
+
+    if (id > 0)
+      output[0] = _local_memory[id - 1];
+  }
+  /// Performs a shuffle operation to move data to the right across the
+  /// work-items, storing the suffix of the group after the shuffle operation.
+  ///
+  /// \tparam ElementsPerWorkItem The number of data elements per work-item.
+  /// \tparam ItemT The work-item identifier type.
+  /// \param item The work-item identifier.
+  /// \param input The input data to be shuffled.
+  /// \param output The array that will store the shuffle result.
+  /// \param group_suffix The suffix of the group after the shuffle.
+  template <int ElementsPerWorkItem, typename ItemT>
+  __syclcompat_inline__ void
+  shuffle_right(const ItemT &item, T (&input)[ElementsPerWorkItem],
```

Comment:
Again, this e.g. maps to https://github.com/intel/llvm/blob/68cacbfed4f54beb5ce61b9963aa6ddca61eebb5/sycl/include/sycl/group_algorithm.hpp#L552

I won't go through all the cases. But I think most of these functions are already implemented in a better way that doesn't just rely on shared memory.

Reply:
Same as above.